src/App.tsx

```diff
@@ -15,33 +15,118 @@ const goodsFromServer: string[] = [
   'Garlic',
 ];
 
-export const App: React.FC = () => (
-  <div className="App">
-    <button type="button">
-      Start
-    </button>
-
-    <button type="button">
-      Sort alphabetically
-    </button>
-
-    <button type="button">
-      Sort by length
-    </button>
-
-    <button type="button">
-      Reverse
-    </button>
-
-    <button type="button">
-      Reset
-    </button>
-
-    <ul className="Goods">
-      <li className="Goods__item">Dumplings</li>
-      <li className="Goods__item">Carrot</li>
-      <li className="Goods__item">Eggs</li>
-      <li className="Goods__item">...</li>
-    </ul>
-  </div>
-);
+const preparedGoods = goodsFromServer.map(name => ({
+  name,
+  id: uuidv4(),
+}));
+
+enum SortType {
+  NONE,
+  ALPHABET,
+  LENGTH,
+}
+
+type Good = {
+  name: string;
+  id: string;
+};
+
+const getReorderedGoods = (
+  goods: Good[],
+  sortType: SortType,
+  isReversed: boolean,
+) => {
+  const visibleGoods = [...goods];
+
+  if (sortType !== SortType.NONE) {
+    visibleGoods.sort((a, b) => {
+      switch (sortType) {
+        case SortType.ALPHABET:
+          return a.name.localeCompare(b.name);
+        case SortType.LENGTH:
+          return a.name.length - b.name.length;
+        default:
+          return 0;
+      }
+    });
+  }
+
+  return isReversed
+    ? visibleGoods.reverse()
+    : visibleGoods;
+};
```

Comment:
You should use `useMemo` hook

Reply:
<img width="492" alt="image" src="https://user-images.githubusercontent.com/32794096/183049690-b7578958-926d-4104-a9e1-0a0a2968aa17.png">

Reply:
I will google about this hook, but if I understand correct about useMemo, I call this function only when one of [isReversed, sortType] states changes.
```
const visibleGoods = useMemo(() => (
    getReorderedGoods(preparedGoods, sortType, isReversed)
  ), [isReversed, sortType]);
```

Reply:
Okay, I see. You moved function out of the component.